tools/manifestfile.py

```diff
@@ -280,29 +362,47 @@ def include(self, manifest_path, top_level=False, **kwargs):
             if manifest_path in self._visited:
                 return
             self._visited.add(manifest_path)
-            if not top_level:
-                self._metadata.append(ManifestMetadata())
-            with open(manifest_path) as f:
-                # Make paths relative to this manifest file while processing it.
-                # Applies to includes and input files.
-                prev_cwd = os.getcwd()
-                os.chdir(os.path.dirname(manifest_path))
-                try:
-                    exec(f.read(), self._manifest_globals(kwargs))
-                except Exception as er:
-                    raise ManifestFileError(
-                        "Error in manifest file: {}: {}".format(manifest_path, er)
-                    )
-                os.chdir(prev_cwd)
-            if not top_level:
+            if is_require:
+                # This include is the result of require("name"), so push a new
+                # package metadata onto the stack.
+                self._metadata.append(ManifestPackageMetadata(is_require=True))
+            try:
+                with open(manifest_path) as f:
+                    # Make paths relative to this manifest file while processing it.
+                    # Applies to includes and input files.
+                    prev_cwd = os.getcwd()
+                    os.chdir(os.path.dirname(manifest_path))
+                    try:
+                        exec(f.read(), self._manifest_globals(kwargs))
+                    finally:
+                        os.chdir(prev_cwd)
+            except ManifestIgnoreException:
+                # e.g. MODE_PYPROJECT and this was a stdlib dependency. No-op.
+                pass
+            except ManifestUsePyPIException as e:
+                # e.g. MODE_PYPROJECT and this was a package from
+                # python-ecosys. Add PyPI dependency instead.
+                self._pypi_dependencies.append(e.pypi_name)
+            except Exception as e:
+                raise ManifestFileError("Error in manifest file: {}: {}".format(manifest_path, e))
+            if is_require:
                 self._metadata.pop()
 
-    def require(self, name, version=None, unix_ffi=False, **kwargs):
+    def require(self, name, version=None, unix_ffi=False, pypi="", **kwargs):
         """
         Require a module by name from micropython-lib.
 
         Optionally specify unix_ffi=True to use a module from the unix-ffi directory.
         """
```

Comment:
maybe add a comment here about the `pypi` argument?

Reply:
Done.